# Command wrappers

```bash

srun -p cpu --time 4:00:00 --ntasks 1 --cpus-per-task 6 --mem 32G --pty /bin/bash

```

## PGS feasibility check across region coverages

We measure the performance of PGS with the coverace of polygenic regions. We compare to the genome-wide case, using GLAD+ genotyped data.

```bash

sbatch --time 3:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R MDDIDP" --output "prs.munge.cojo.MDDIDP.$(date +%Y%m%d).out.txt"
sbatch --time 3:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R BIPO03" --output "prs.munge.cojo.BIPO03.$(date +%Y%m%d).out.txt"
sbatch --time 3:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R ADHD06" --output "prs.munge.cojo.ADHD06.$(date +%Y%m%d).out.txt"
sbatch --time 3:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R SCHI06" --output "prs.munge.cojo.SCHI06.$(date +%Y%m%d).out.txt"
sbatch --time 3:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R SMOK11" --output "prs.munge.cojo.SMOK11.$(date +%Y%m%d).out.txt"


sbatch --time 2-00:00:00 --partition cpu --job-name="MDDIDP" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R MDDIDP" --output "sbayesrc.tidyimpute.MDDIDP.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="BIPO03" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R BIPO03" --output "sbayesrc.tidyimpute.BIPO03.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="ADHD06" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R ADHD06" --output "sbayesrc.tidyimpute.ADHD06.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="SCHI06" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R SCHI06" --output "sbayesrc.tidyimpute.SCHI06.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="SMOK11" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R SMOK11" --output "sbayesrc.tidyimpute.SMOK11.$(date +%Y%m%d).out.txt"


sbatch --time 2-00:00:00 --partition cpu --job-name="MDDIDP" --ntasks 1 --cpus-per-task 8 --mem 32G --wrap="Rscript ../manual_routines/prs-check-gladp.R MDDIDP" --output "sbayesrc.score.MDDIDP.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="BIPO03" --ntasks 1 --cpus-per-task 8 --mem 32G --wrap="Rscript ../manual_routines/prs-check-gladp.R BIPO03" --output "sbayesrc.score.BIPO03.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="ADHD06" --ntasks 1 --cpus-per-task 8 --mem 32G --wrap="Rscript ../manual_routines/prs-check-gladp.R ADHD06" --output "sbayesrc.score.ADHD06.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="SCHI06" --ntasks 1 --cpus-per-task 8 --mem 32G --wrap="Rscript ../manual_routines/prs-check-gladp.R SCHI06" --output "sbayesrc.score.SCHI06.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="SMOK11" --ntasks 1 --cpus-per-task 8 --mem 32G --wrap="Rscript ../manual_routines/prs-check-gladp.R SMOK11" --output "sbayesrc.score.SMOK11.$(date +%Y%m%d).out.txt"


sbatch --time 4:00:00 --partition cpu --job-name="scoring" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R" --output "sbayesrc.scoring.baseline.$(date +%Y%m%d).out.txt"
#needs plink2
sbatch --time 4:00:00 --partition cpu --job-name="scoring" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript ../manual_routines/prs-check-gladp.R" --output "sbayesrc.scoring.allPolygenic.$(date +%Y%m%d).out.txt"

```

## ONT human variation workflow

### Installation and configuration

```bash

export JAVA_HOME='/scratch/prj/ppn_tng/software/java/jdk-24.0.1'
export JAVA_CMD='/scratch/prj/ppn_tng/software/java/jdk-24.0.1/bin/java'

export NEXTFLOW_CMD='/scratch/prj/ppn_tng/software/nextflow/nextflow-25.04.4-dist'

$NEXTFLOW_CMD pull epi2me-labs/wf-human-variation

$NEXTFLOW_CMD run epi2me-labs/wf-human-variation --help

```

## Singularity

Install singularity. Singularity can't build containers on MacOS currently. You can use a virtual machine (f.ex. using VirtualBox) to run Linux and then build Singularity that can build containers from definitions.
https://github.com/sylabs/singularity/blob/main/INSTALL.md
https://docs.sylabs.io/guides/3.0/user-guide/installation.html


Install prerequisites
```bash
sudo pacman -Syu
sudo pacman -Sy autoconf automake cryptsetup crun fuse2fs git fuse3 libseccomp libtool pkgconf shadow squashfs-tools wget zlib conmon go base-devel

#base-devel includes gcc and make

```

After restart, this is the rouine of the installation.
```bash
#in projects directory
git clone --recurse-submodules https://github.com/sylabs/singularity.git
cd singularity
#git submodule update --init #only if you don't already have the submodules
./mconfig
make -C builddir
sudo make -C builddir install

```


# Building a mosdepth container/image

```bash



#this was run historically - using a slimmed down version of mosdepth_c2.def to identify build errors
#sudo singularity build --sandbox mosdepth_c2 /media/sf_mylinux_shared/mosdepth_c2.def

sudo singularity build mosdepth_c2.sif /media/sf_mylinux_shared/mosdepth_c2.def
#singularity build --fakeroot mosdepth_c2.sif /media/sf_mylinux_shared/mosdepth_c2.def #this did not work

#test of using shared folder for cache and tempdir - THIS DOES NOT WORK CURRENTLY!
#if required - set the group permission for the user (q)
sudo usermod -aG vboxsf q
#sudo -E required in archlinux according to https://github.com/apptainer/singularity/issues/5791
#enable symlinks in the shared folder (virtualbox mylinux and share mylinux_shared)
VBoxManage setextradata mylinux VBoxInternal2/SharedFoldersEnableSymlinksCreate/mylinux_shared 1
export SINGULARITY_TMPDIR="/media/sf_mylinux_shared/singularity_tmp"
export SINGULARITY_CACHEDIR="/media/sf_mylinux_shared/singularity_cache"
sudo -E singularity build prada_c2.sif prada_c2.def #as root, in /media/sf_mylinux_shared/singularity_tmp
#singularity build --fakeroot prada_c2.sif prada_c2.def #does not work still

#clean environment
sudo pacman -Sc
sudo pacman -Scc
singularity cache clean

#running in larger vmachine
export SINGULARITY_TMPDIR="/home/q/Documents/work/sing_temp"
export SINGULARITY_CACHEDIR="/home/q/Documents/work/sing_cache"
sudo -E singularity build prada_c2.sif /media/sf_mylinux_shared/prada_c2.def
#this works. not sure if it is the change in disk or the use of -E.

```


#Manual mosdepths runs

```bash

#interactive test
#singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --fast-mode barcode01 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-nobedtest/output/barcode01/barcode01.haplotagged.bam

#my nogtube-nobed test run

sbatch --time 24:00:00 --partition cpu --job-name="md01" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode01 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-nobedtest/output/barcode01/barcode01.haplotagged.bam" --output "mdepth.barcode01.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md02" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode02 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-nobedtest/output/barcode02/barcode02.haplotagged.bam" --output "mdepth.barcode02.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md03" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode03 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-nobedtest/output/barcode03/barcode03.haplotagged.bam" --output "mdepth.barcode03.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md04" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode04 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-nobedtest/output/barcode04/barcode04.haplotagged.bam" --output "mdepth.barcode04.$(date +%Y%m%d).out.txt"


#Bhakti's nogtube run

sbatch --time 24:00:00 --partition cpu --job-name="md01" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode01 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-b1/output/barcode01/barcode01.haplotagged.bam" --output "mdepth.barcode01.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md02" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode02 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-b2/output/barcode02/barcode02.haplotagged.bam" --output "mdepth.barcode02.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md03" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode03 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-b3/output/barcode03/barcode03.haplotagged.bam" --output "mdepth.barcode03.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md04" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode04 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-nogtube-b4/output/barcode04/barcode04.haplotagged.bam" --output "mdepth.barcode04.$(date +%Y%m%d).out.txt"

#Bhakti's gtube run

sbatch --time 24:00:00 --partition cpu --job-name="md01" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode01 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-gtube-b1/output/barcode01/barcode01.haplotagged.bam" --output "mdepth.barcode01.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md02" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode02 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-gtube-b2/output/barcode02/barcode02.haplotagged.bam" --output "mdepth.barcode02.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md03" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode03 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-gtube-b3/output/barcode03/barcode03.haplotagged.bam" --output "mdepth.barcode03.$(date +%Y%m%d).out.txt"

sbatch --time 24:00:00 --partition cpu --job-name="md04" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="singularity run --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz ~/project/prada_jz/venv/mosdepth_c2.sif -t 6 --by /scratch/prj/sgdp_nanopore/Projects/prada_jz/data/bed/pgx_cnv.grch38.5k.2p1percent.bed --fast-mode barcode04 /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/pgx/pilot2/p2-gtube-b4/output/barcode04/barcode04.haplotagged.bam" --output "mdepth.barcode04.$(date +%Y%m%d).out.txt"


```


#Creating IGV graphs


```bash

gunzip -c barcode01.regions.bed.gz | awk '{print $1, $2, $3, $5}' > barcode01.regions.bedgraph
gunzip -c barcode01.per-base.bed.gz | awk '{print $1, $2, $3, $4}' > barcode01.per-base.bedgraph

gunzip -c barcode02.regions.bed.gz | awk '{print $1, $2, $3, $5}' > barcode02.regions.bedgraph
gunzip -c barcode02.per-base.bed.gz | awk '{print $1, $2, $3, $4}' > barcode02.per-base.bedgraph

gunzip -c barcode03.regions.bed.gz | awk '{print $1, $2, $3, $5}' > barcode03.regions.bedgraph
gunzip -c barcode03.per-base.bed.gz | awk '{print $1, $2, $3, $4}' > barcode03.per-base.bedgraph

gunzip -c barcode04.regions.bed.gz | awk '{print $1, $2, $3, $5}' > barcode04.regions.bedgraph
gunzip -c barcode04.per-base.bed.gz | awk '{print $1, $2, $3, $4}' > barcode04.per-base.bedgraph

#interactive
#test of automated igvtools conversion to tdf
singularity shell --bind /scratch/prj/sgdp_nanopore/Projects/prada_jz /scratch/prj/sgdp_nanopore/Projects/prada_jz/venv/prada_c2.sif

igvtools toTDF /scratch/prj/sgdp_nanopore/Projects/prada_jz/work/mosdepth/pilot2/p2-nogtube-nobedtest/barcode01.regions.bedgraph barcode01.regions.tdf hg38 #this throws an error


```


#Analysing existing pilot runs on Create

Run in work/pradaApp

## Interactive version

```{R}


library(prada)
library(data.table)

projectFolderPath<-"/scratch/prj/sgdp_nanopore/Projects/prada_jz"
pradaObj<-PradaClass()

pradaObj$addAnalysisSetting(settingLabel = "p2-gtube",folderPathAnalysisSequencingRaw = file.path(projectFolderPath, "data/ont_raw/pilot2/Gtube/20250724_1536_3B_PAW94949_16dd4442"),folderPathAnalysisOutputRaw = file.path(projectFolderPath, "work/pgx/pilot2/p2-gtube"),folderPathDepthAnalysisOutputRaw = file.path(projectFolderPath,"work/mosdepth/pilot2/p2-gtube"))
pradaObj$addAnalysisSetting(settingLabel = "p2-nogtube",folderPathAnalysisSequencingRaw = file.path(projectFolderPath,"data/ont_raw/pilot2/No_Gtube/20250724_1536_3C_PAY03690_092497cc"),folderPathAnalysisOutputRaw = file.path(projectFolderPath,"work/pgx/pilot2/p2-nogtube"),folderPathDepthAnalysisOutputRaw = file.path(projectFolderPath,"work/mosdepth/pilot2/p2-nogtube"))
pradaObj$collectAnalysisCallData("p2-gtube")
pradaObj$collectAnalysisCallData("p2-nogtube")
pradaObj$collectAnalysisDepthData("p2-gtube")
pradaObj$collectAnalysisDepthData("p2-nogtube")
pradaObj$computeDepthDataStatistics(filePathBed <- file.path(projectFolderPath,"data/bed/pgx_cnv.grch38.5k.2p1percent.bed"),filePathApplicationCoverageRegions = file.path(projectFolderPath,"data/applicationCoverageRegionsAsOfPilot2.tsv"))
pradaObj$computeCallStatistics(filePathApplicationCoverageRegions = file.path(projectFolderPath,"data/applicationCoverageRegionsAsOfPilot2.tsv"))
pradaObj$printData()


#plotTable<-as.data.frame(matrix(NA,0,0))

sampleMeta<-data.table::fread("sampleMeta.tsv")
sampleMeta[,sampleLabel:=paste0(analysis,"_",barcode)]

if(nrow(sampleMeta)>0){
      for(iSample in 1:nrow(sampleMeta)){
        #iSample<-1
        cAnalysisLabel<-sampleMeta[iSample,c("analysis")]
        cBarcode<-sampleMeta[iSample,c("barcode")]
        cUniqueSampleLabel <- paste0(cAnalysisLabel,"_",cBarcode)
        
        cStats<-data.table::fread(paste0("sequencingStatsOriginalRegionsTable_",cUniqueSampleLabel,".tsv"))
        cPgx<-data.table::fread(paste0("pgxCallsAggCustom_",cUniqueSampleLabel,".tsv"))
        cPgx[,pgxVariantRatio:=nvariants_mean/(nvariants_mean+nmissingpos_mean)]
        
        sampleMeta[sampleLabel==eval(cUniqueSampleLabel),c("nCalledPgx")]<-nrow(cPgx[!is.na(diplotype_score_top),])
        sampleMeta[sampleLabel==eval(cUniqueSampleLabel),c("meanPgxVariantRatio")]<-mean(cPgx$pgxVariantRatio,na.rm=T)
        
      }
        
}


plot(sampleMeta$sdepth_q050_bed,sampleMeta$nvarq_region)

plot(sampleMeta$sdepth_q050_bed,sampleMeta$meanPgxVariantRatio)

plot(sampleMeta$sdepth_q050_bed,sampleMeta$nCalledPgx)

plot(sampleMeta$sdepth_q050_bed,sampleMeta$vcallacc_q050_region)


```


```bash

sbatch --time 1:00:00 --partition cpu --job-name="pap2" --ntasks 1 --cpus-per-task 6 --mem 16G --wrap="Rscript /scratch/prj/sgdp_nanopore/Projects/prada_jz/src-r/pradaAppPilot2.R" --output "pradaAppPilot2.$(date +%Y%m%d).out.txt"


```



